library/HTMLPurifier/HTMLModule/MathML.php

```diff
@@ -0,0 +1,2254 @@
+<?php
+
+/**
+ * MathML 3 specification.
+ */
+class HTMLPurifier_HTMLModule_MathML extends HTMLPurifier_HTMLModule
+{
+    /**
+     * @type string
+     */
+    public $name = 'MathML';
+
+    // Prefix in case MathML is imported
+    private $mathml_prefix = 'm';
+
+    /**
+     * @param HTMLPurifier_Config $config
+     */
+    public function setup($config)
+    {
+
+        // Normalize whitespace inside text elements as per MathML spec 2.1.7
+        $this->info_injector[] = new HTMLPurifier_Injector_MathMLSpaceNormalize();
+
+        /*****************************************************************
+         * Meta variables
+         * Used in this file to simplify code and help adapt the DTD
+         *****************************************************************/
+
+        // The elements inside <math> are not supposed to be outside, so they
+        // can't be neither block nor inline:
+        $default_display = false;
+
+        // This array will contain all the necessary DTD entities used
+        // throughout the MathML DTD, in order to avoid collisions and allow
+        // for special characters ('.', '-') in entity names
+        $E = array();
+
+        // Prefix used for xlink attrs; is not specified by the MathML DTD
+        $E['XLINK.prefix'] = 'xlink';
+
+        $proprietary_att_wrs = array(
```

Comment:
What are these attributes?

Reply:
`dsi:*` and `wrs:*` are a set of proprietary styling attributes which are pretty common in many MathML documents.
In fact, they appear in the standard W3C testing suite (see e.g. https://www.w3.org/Math/testsuite/mml2-testsuite/Presentation/DynamicExpressions/maction/mactionBhigh1.xml) which we use to validate the MathML module in HTMLPurifier.
Even thought they are not part of the standard, we believe them to be important enough to be validated just like the rest of the attributes.